File: TelerikUIExamples/TelerikUIExamples/IndicatorsTableView.m
//
//  IndicatorsTableViewViewController.m
//  TelerikUIExamples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

#import "IndicatorsTableView.h"
#import "IndicatorsViewController.h"
#import "OptionInfo.h"

NSString *const kCellIdentifier = @"indicatorCell";
@implementation IndicatorsTableView

- (void)viewDidLoad
{
    [super viewDidLoad];
    _table = [[UITableView alloc] initWithFrame:self.view.bounds];
    _table.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
    _table.dataSource = self;
    _table.delegate = self;
    _table.allowsMultipleSelection = YES;
    _table.backgroundColor = [UIColor whiteColor];
    [self.view addSubview:_table];
    
    NSIndexPath *trendlinePath = [NSIndexPath indexPathForRow:_example.selectedTrendline inSection:0];
    NSIndexPath *indicatorPath = [NSIndexPath indexPathForRow:_example.selectedIndicator inSection:1];
    [_table selectRowAtIndexPath:trendlinePath animated:NO scrollPosition:UITableViewScrollPositionNone];
    [_table selectRowAtIndexPath:indicatorPath animated:NO scrollPosition:UITableViewScrollPositionNone];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 2;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    return 40;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if (section == 0) {
        return _example.trendlines.count;
    } else {
        return _example.indicators.count;
    }
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if (section == 0) {
        return @"Trendlines";
    } else {
        return @"Indicators";
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSInteger section = indexPath.section;
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellIdentifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:kCellIdentifier];
    }
    
    OptionInfo *info;
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    if (section == 0) {
        info = _example.trendlines[indexPath.row];
        if (indexPath.row == _example.selectedTrendline) {
            cell.accessoryType = UITableViewCellAccessoryCheckmark;
        } else {
            cell.accessoryType = UITableViewCellAccessoryNone;
        }
    } else {
        info = _example.indicators[indexPath.row];
        if (indexPath.row == _example.selectedIndicator) {
            cell.accessoryType = UITableViewCellAccessoryCheckmark;
        } else {
            cell.accessoryType = UITableViewCellAccessoryNone;
        }
    }
    
    cell.textLabel.text = info.optionText;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSInteger section = indexPath.section;
    NSInteger selectedRow = indexPath.row;
    OptionInfo *info;
    if (section == 0) {
        NSIndexPath *currentPath = [NSIndexPath indexPathForRow:_example.selectedTrendline inSection:section];
        [tableView deselectRowAtIndexPath:currentPath animated:NO];
        _example.selectedTrendline = selectedRow;
        info = _example.trendlines[selectedRow];
    } else {
        NSIndexPath *currentPath = [NSIndexPath indexPathForRow:_example.selectedIndicator inSection:section];
        [tableView deselectRowAtIndexPath:currentPath animated:NO];
        _example.selectedIndicator = selectedRow;
        info = _example.indicators[selectedRow];
    }
    
    IMP indicatorImp = [_example methodForSelector:info.selector];
    void (*indicatorFunc)(id, SEL, OptionInfo *) = (void *)indicatorImp;
    indicatorFunc(_example, info.selector, info);
    
    [tableView reloadData];
}

- (NSIndexPath *)tableView:(UITableView *)tableView willDeselectRowAtIndexPath:(NSIndexPath *)indexPath
{
    return nil;
}


@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/IndicatorsTableView.swift
//
//  IndicatorsTableView.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation

class IndicatorsTableView : UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    var example: Indicators?
    var table = UITableView()
    let kCellIdentifier: NSString = "indicatorCell"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        table.frame = self.view.bounds
        table.autoresizingMask = UIViewAutoresizing.FlexibleHeight | UIViewAutoresizing.FlexibleWidth
        table.dataSource = self
        table.delegate = self
        table.allowsMultipleSelection = true
        table.registerClass(UITableViewCell.self, forCellReuseIdentifier: kCellIdentifier)
        self.view.addSubview(table)
        
        let trendlinePath = NSIndexPath(forRow: example!.selectedTrendline, inSection: 0)
        let indicatorPath = NSIndexPath(forRow: example!.selectedIndicator, inSection: 1)
        table.selectRowAtIndexPath(trendlinePath, animated: false, scrollPosition: UITableViewScrollPosition.None)
        table.selectRowAtIndexPath(indicatorPath, animated: false, scrollPosition: UITableViewScrollPosition.None)
    }
    
    //MARK: - UITableViewDataSource
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 2;
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return example!.trendlines.count
        }
        else {
            return example!.indicators.count
        }
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var info: OptionInfo
        let cell = tableView.dequeueReusableCellWithIdentifier(kCellIdentifier) as UITableViewCell!
        let section = indexPath.section as NSInteger
        if section == 0 {
            info = example!.trendlines[indexPath.row] as OptionInfo
            if indexPath.row == example!.selectedTrendline {
                cell.accessoryType = UITableViewCellAccessoryType.Checkmark
            }
            else {
                cell.accessoryType = UITableViewCellAccessoryType.None
            }
        }
        else {
            info = example!.indicators[indexPath.row] as OptionInfo
            if indexPath.row == example!.selectedIndicator {
                cell.accessoryType = UITableViewCellAccessoryType.Checkmark
            }
            else {
                cell.accessoryType = UITableViewCellAccessoryType.None
            }
        }
        
        cell.selectionStyle = UITableViewCellSelectionStyle.None
        cell.textLabel!.text = info.optionText
        
        return cell
    }
    
    //MARK: - UITableViewDelegate
    
    func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 40
    }

    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String {
        if section == 0 {
            return "Trendlines"
        }
        else {
            return "Indicators"
        }
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let section = indexPath.section
        let selectedRow = indexPath.row
        var info: OptionInfo
        if section == 0 {
            let currentPath = NSIndexPath(forRow: example!.selectedTrendline, inSection: section)
            tableView.deselectRowAtIndexPath(currentPath, animated: false)
            example!.selectedTrendline = selectedRow
            info = example!.trendlines[selectedRow] as OptionInfo
        }
        else {
            let currentPath = NSIndexPath(forRow: example!.selectedTrendline, inSection: section)
            tableView.deselectRowAtIndexPath(currentPath, animated: false)
            example!.selectedIndicator = selectedRow
            info = example!.indicators[selectedRow] as OptionInfo
        }

        info.selector!()
        
        tableView.reloadData()
    }

    func tableView(tableView: UITableView, willDeselectRowAtIndexPath indexPath: NSIndexPath) -> NSIndexPath? {
        return nil
    }
}